File: lucene-directory/src/main/java/jetbrains/exodus/lucene/DebugExodusDirectory.java
/**
 * Copyright 2010 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.lucene;

import jetbrains.exodus.env.ContextualEnvironment;
import jetbrains.exodus.env.StoreConfig;
import jetbrains.exodus.vfs.VfsConfig;
import org.apache.lucene.index.IndexFileNames;
import org.apache.lucene.store.*;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.Arrays;
import java.util.Collection;

public class DebugExodusDirectory extends Directory {

    private final ExodusDirectory directory;
    private final RAMDirectory debugDirectory;

    public DebugExodusDirectory(@NotNull final ContextualEnvironment env) {
        this(env, StoreConfig.WITH_DUPLICATES);
    }

    public DebugExodusDirectory(@NotNull final ContextualEnvironment env,
                                @NotNull final StoreConfig contentsStoreConfig) {
        directory = new ExodusDirectory(env, contentsStoreConfig);
        debugDirectory = new RAMDirectory();
    }

    public DebugExodusDirectory(@NotNull final ContextualEnvironment env,
                                @NotNull final VfsConfig vfsConfig,
                                @NotNull final StoreConfig contentsStoreConfig) {
        directory = new ExodusDirectory(env, vfsConfig, contentsStoreConfig);
        debugDirectory = new RAMDirectory();
    }

    @Override
    public String[] listAll() {
        final String[] result = directory.listAll();
        final String[] debugResult = debugDirectory.listAll();
        if (result.length != debugResult.length) {
            throwDebugMismatch();
        }
        Arrays.sort(result);
        Arrays.sort(debugResult);
        int i = 0;
        for (String file : debugResult) {
            if (!file.equals(result[i++])) {
                throwDebugMismatch();
            }
        }
        return result;
    }

    @Override
    public void deleteFile(String name) throws IOException {
        directory.deleteFile(name);
        debugDirectory.deleteFile(name);
    }

    @Override
    public long fileLength(String name) throws IOException {
        final long result = directory.fileLength(name);
        if (result != debugDirectory.fileLength(name)) {
            throwDebugMismatch();
        }
        return result;
    }

    @Override
    public IndexOutput createOutput(String name, IOContext context) throws IOException {
        return new DebugIndexOutput(name, context);
    }

    @Override
    public IndexOutput createTempOutput(String prefix, String suffix, IOContext context) throws IOException {
        return createOutput(IndexFileNames.segmentFileName(prefix, suffix + '_' + directory.nextTicks(), "tmp"), context);
    }

    @Override
    public void sync(Collection<String> names) throws IOException {
        directory.sync(names);
        debugDirectory.sync(names);
    }

    @Override
    public void rename(String source, String dest) throws IOException {
        directory.rename(source, dest);
        debugDirectory.rename(source, dest);
    }

    @Override
    public void syncMetaData() throws IOException {
        directory.syncMetaData();
        debugDirectory.syncMetaData();
    }

    @Override
    public IndexInput openInput(String name, IOContext context) throws IOException {
        return new DebugIndexInput(name, context);
    }

    @Override
    public Lock obtainLock(String name) throws IOException {
        return debugDirectory.obtainLock(name);
    }

    @Override
    public void close() {
        directory.close();
        debugDirectory.close();
    }

    private static void throwDebugMismatch() {
        throw new RuntimeException("Debug directory mismatch");
    }

    private class DebugIndexOutput extends IndexOutput {

        private final IndexOutput output;
        private final IndexOutput debugOutput;

        private DebugIndexOutput(String name, IOContext context) throws IOException {
            super("DebugIndexOutput for " + name, name);
            output = directory.createOutput(name, context);
            debugOutput = debugDirectory.createOutput(name, context);
        }

        @Override
        public void close() throws IOException {
            output.close();
            debugOutput.close();
        }

        @Override
        public long getFilePointer() {
            final long result = output.getFilePointer();
            if (result != debugOutput.getFilePointer()) {
                throwDebugMismatch();
            }
            return result;
        }

        @Override
        public long getChecksum() throws IOException {
            final long result = output.getChecksum();
            if (result != debugOutput.getChecksum()) {
                throwDebugMismatch();
            }
            return result;
        }

        @Override
        public void writeByte(byte b) throws IOException {
            output.writeByte(b);
            debugOutput.writeByte(b);
            getFilePointer();
        }

        @Override
        public void writeBytes(byte[] b, int offset, int length) throws IOException {
            output.writeBytes(b, offset, length);
            debugOutput.writeBytes(b, offset, length);
            getFilePointer();
        }
    }

    @SuppressWarnings("CloneableClassInSecureContext")
    private class DebugIndexInput extends IndexInput {

        private String name;
        private IndexInput input;
        private IndexInput debugInput;

        private DebugIndexInput(String name, IOContext context) throws IOException {
            this(name, context, 0L);
        }

        private DebugIndexInput(String name, IOContext context, long position) throws IOException {
            super("DebugIndexInput for " + name);
            this.name = name;
            input = directory.openInput(name, context);
            debugInput = debugDirectory.openInput(name, context);
            if (position > 0) {
                input.seek(position);
                debugInput.seek(position);
            }
        }

        @Override
        public void close() throws IOException {
            input.close();
            debugInput.close();
        }

        @Override
        public long getFilePointer() {
            final long result = input.getFilePointer();
            if (result != debugInput.getFilePointer()) {
                throwDebugMismatch();
            }
            return result;
        }

        @Override
        public void seek(long pos) throws IOException {
            input.seek(pos);
            debugInput.seek(pos);
            if (input.getFilePointer() != debugInput.getFilePointer()) {
                throwDebugMismatch();
            }
        }

        @Override
        public long length() {
            final long result = input.length();
            if (result != debugInput.length()) {
                throwDebugMismatch();
            }
            return result;
        }

        @Override
        public byte readByte() throws IOException {
            final byte result = input.readByte();
            final byte debugByte = debugInput.readByte();
            if (result != debugByte) {
                throwDebugMismatch();
            }
            return result;
        }

        @Override
        public void readBytes(byte[] b, int offset, int len) throws IOException {
            final long before = getFilePointer();
            input.readBytes(b, offset, len);
            final byte[] bytes = new byte[len];
            debugInput.readBytes(bytes, 0, len);
            final long after = getFilePointer();
            for (int i = 0; i < (int) (after - before); ++i) {
                if (bytes[i] != b[offset + i]) {
                    throwDebugMismatch();
                }
            }
        }

        @Override
        public final IndexInput clone() {
            final DebugIndexInput result = (DebugIndexInput) super.clone();
            result.name = name;
            result.input = input.clone();
            result.debugInput = debugInput.clone();
            return result;
        }

        @Override
        public IndexInput slice(String sliceDescription, final long offset, final long length) throws IOException {
            final DebugIndexInput result = (DebugIndexInput) super.clone();
            result.name = name;
            result.input = input.slice(sliceDescription, offset, length);
            result.debugInput = debugInput.slice(sliceDescription, offset, length);
            return result;
        }
    }
}

File: lucene-directory/src/main/kotlin/jetbrains/exodus/lucene/DebugExodusDirectory.kt
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.lucene

import jetbrains.exodus.env.ContextualEnvironment
import jetbrains.exodus.env.StoreConfig
import org.apache.lucene.index.IndexFileNames
import org.apache.lucene.store.*

class DebugExodusDirectory : Directory {

    private val directory: ExodusDirectory
    private val debugDirectory: RAMDirectory

    @JvmOverloads constructor(env: ContextualEnvironment,
                              contentsStoreConfig: StoreConfig = StoreConfig.WITH_DUPLICATES,
                              directoryConfig: ExodusDirectoryConfig = ExodusDirectoryConfig()) {
        directory = ExodusDirectory(env, contentsStoreConfig, directoryConfig)
        debugDirectory = RAMDirectory()
    }

    override fun listAll(): Array<String> = debugDirectory.listAll()

    override fun deleteFile(name: String) {
        directory.deleteFile(name)
        debugDirectory.deleteFile(name)
    }

    override fun fileLength(name: String): Long {
        val result = directory.fileLength(name)
        if (result != debugDirectory.fileLength(name)) {
            throwDebugMismatch()
        }
        return result
    }

    override fun createOutput(name: String, context: IOContext): IndexOutput {
        return DebugIndexOutput(name, context)
    }

    override fun createTempOutput(prefix: String, suffix: String, context: IOContext): IndexOutput {
        return createOutput(IndexFileNames.segmentFileName(prefix, suffix + '_' + directory.nextTicks(), "tmp"), context)
    }

    override fun sync(names: Collection<String>) {
        directory.sync(names)
        debugDirectory.sync(names)
    }

    override fun rename(source: String, dest: String) {
        directory.rename(source, dest)
        debugDirectory.rename(source, dest)
    }

    override fun syncMetaData() {
        directory.syncMetaData()
        debugDirectory.syncMetaData()
    }

    override fun openInput(name: String, context: IOContext): IndexInput {
        return DebugIndexInput(name, context)
    }

    override fun obtainLock(name: String): Lock {
        return debugDirectory.obtainLock(name)
    }

    override fun getPendingDeletions(): MutableSet<String> {
        return directory.pendingDeletions.also {
            val debugDeletions = debugDirectory.pendingDeletions
            if (it.size != debugDeletions.size || !it.containsAll(debugDeletions)) {
                throwDebugMismatch()
            }
        }
    }

    override fun close() {
        directory.close()
        debugDirectory.close()
    }

    private fun throwDebugMismatch() {
        throw RuntimeException("Debug directory mismatch")
    }

    private inner class DebugIndexOutput(name: String, context: IOContext) :
        IndexOutput("DebugIndexOutput[$name]", name) {

        private val output: IndexOutput = directory.createOutput(name, context)
        private val debugOutput: IndexOutput = debugDirectory.createOutput(name, context)

        override fun close() {
            output.close()
            debugOutput.close()
        }

        override fun getFilePointer(): Long {
            val result = output.filePointer
            if (result != debugOutput.filePointer) {
                throwDebugMismatch()
            }
            return result
        }

        override fun getChecksum(): Long {
            val result = output.checksum
            if (result != debugOutput.checksum) {
                throwDebugMismatch()
            }
            return result
        }

        override fun writeByte(b: Byte) {
            output.writeByte(b)
            debugOutput.writeByte(b)
            filePointer
        }

        override fun writeBytes(b: ByteArray, offset: Int, length: Int) {
            output.writeBytes(b, offset, length)
            debugOutput.writeBytes(b, offset, length)
            filePointer
        }
    }

    private inner class DebugIndexInput(private var name: String, context: IOContext, position: Long = 0L) :
        IndexInput("DebugIndexInput[$name]") {

        private var input = directory.openInput(name, context)
        private var debugInput = debugDirectory.openInput(name, context)

        init {
            input = directory.openInput(name, context)
            debugInput = debugDirectory.openInput(name, context)
            if (position > 0) {
                input.seek(position)
                debugInput.seek(position)
            }
        }

        override fun close() {
            input.close()
            debugInput.close()
        }

        override fun getFilePointer(): Long {
            val result = input.filePointer
            if (result != debugInput.filePointer) {
                throwDebugMismatch()
            }
            return result
        }

        override fun seek(pos: Long) {
            input.seek(pos)
            debugInput.seek(pos)
            if (input.filePointer != debugInput.filePointer) {
                throwDebugMismatch()
            }
        }

        override fun length(): Long {
            val result = input.length()
            if (result != debugInput.length()) {
                throwDebugMismatch()
            }
            return result
        }

        override fun readByte(): Byte {
            val result = input.readByte()
            val debugByte = debugInput.readByte()
            if (result != debugByte) {
                throwDebugMismatch()
            }
            return result
        }

        override fun readBytes(b: ByteArray, offset: Int, len: Int) {
            val before = filePointer
            input.readBytes(b, offset, len)
            val bytes = ByteArray(len)
            debugInput.readBytes(bytes, 0, len)
            val after = filePointer
            for (i in 0 until (after - before).toInt()) {
                if (bytes[i] != b[offset + i]) {
                    throwDebugMismatch()
                }
            }
        }

        override fun clone(): IndexInput {
            val result = super.clone() as DebugIndexInput
            result.name = name
            result.input = input.clone()
            result.debugInput = debugInput.clone()
            return result
        }

        override fun slice(sliceDescription: String, offset: Long, length: Long): IndexInput {
            val result = super.clone() as DebugIndexInput
            result.name = name
            result.input = input.slice(sliceDescription, offset, length)
            result.debugInput = debugInput.slice(sliceDescription, offset, length)
            return result
        }
    }
}
